### PR TITLE
build: update dev deps

### LINKED ISSUE #236: Update all DEV dependencies
Some of the DEV dependencies are quite tricky, and have interdependencies. I am creating this issue, in the hope that @RebeccaStevens will be interested to help update those, as she was the one who had them set up in the first place. I got a bit stuck trying to bring them all up to date.

Hi @RebeccaStevens! Would you, please help us out, to update all DEV dependencies! 🙏 Thank you!!! 🫶

### PR DESCRIPTION
**Pull request by RebeccaStevens**
fix #236

It looks like a lot but most of it is just lining updates.
Look at just the first commit to see what's actually changed.

### CODE REVIEW

**Comment by RebeccaStevens** [comment]
Looks like the testing update I made doesn't work in node 16.
How would you like to resolve this? Drop support for node 16? or should I see if I can get it working.

**Comment by vitaly-t** [comment]
> Looks like the testing update I made doesn't work in node 16. How would you like to resolve this? Drop support for node 16? or should I see if I can get it working.

What's the minimum version that does work? is it v18? Technically, this is just for testing and assembling a release, it is not a functional limitation, so I am quite ok to up min node version to 18, if needed ;)

**Comment by vitaly-t** [comment]
Other than that, as I went through the scope of changes - that looks huge! I'm particularly intrigued by adding `type` everywhere. Can you tell me, please, why it became needed?

**Comment by RebeccaStevens** [comment]
> > Looks like the testing update I made doesn't work in node 16. How would you like to resolve this? Drop support for node 16? or should I see if I can get it working.
> 
> What's the minimum version that does work? is it v18? Technically, this is just for testing and assembling a release, it is not a functional limitation, so I am quite ok to up min node version to 18, if needed ;)

Looks like v18 works

File: tsconfig.base.json
Comment (vitaly-t) [inline tsconfig.base.json:1]: I don't think we can have any comments inside a JSON file.

File: package.json [hunk 1]
```diff
@@ -39,65 +39,73 @@
...
-    "typedoc": "0.24.8",
-    "typescript": "5.1.6"
+    "tsx": "4.16.5",
+    "typedoc": "0.26.5",
+    "typescript": "5.5.4",
+    "typescript-eslint": "8.0.0"
   },
   "packageManager": "yarn@1.22.22",
   "engines": {
     "node": ">=16"
```
Comment (vitaly-t) [inline package.json:109]: We will need to up the Node version to make tests work.

**Review by vitaly-t — CHANGES_REQUESTED** [review]
Couple things  - comments in JSON, plus the Node version.

**Comment by RebeccaStevens** [comment]
> addition of `type` everywhere. Can you tell me, please, why it became needed?

First thing to note, I replaced ts-node with tsx.
When running the tests, tsx will strip out all the imports/exports marked as type imports/exports. If they weren't stripped out we'd get a runtime error when trying to import the type at runtime.
All though it's possible to not have to explicitly mark each type import, and continue to run things as we did with ts-node, explicitly marking them comes with additional improvements.
Additionally, I setup the the eslint rule `@typescript-eslint/consistent-type-imports` and `'@typescript-eslint/no-import-type-side-effects` to enforce this (they have fixers so you don't need to worry about explicitly adding `type` when writing code).

File: tsconfig.base.json
Comment (RebeccaStevens) [reply tsconfig.base.json:1]: tsconfig json files are actually jsonc files, despite the extension.

**Comment by vitaly-t** [comment]
> > addition of `type` everywhere. Can you tell me, please, why it became needed?
> 
> First thing to note, I replaced ts-node with tsx. When running the tests, tsx will strip out all the imports/exports marked as type imports/exports. If they weren't stripped out we'd get a runtime error when trying to import the type at runtime. All though it's possible to not have to explicitly mark each type import, and continue to run things as we did with ts-node, explicitly marking them comes with additional improvements. Additionally, I setup the the eslint rule `@typescript-eslint/consistent-type-imports` and `'@typescript-eslint/no-import-type-side-effects` to enforce this (they have fixers so you don't need to worry about explicitly adding `type` when writing code).

Thank you for the clarification! 🤗

**Review by vitaly-t — APPROVED** [review]

**Comment by vitaly-t** [comment]
@RebeccaStevens 

LGTM, thank you for this incredible work, I would not have been able to make it myself!

Leaving it for you to merge whenever you think it is ready :wink:

**Pull request merged**